### PR TITLE
[SPARK-47418][SQL] Add hand-crafted implementations for lowercase unicode-aware contains, startsWith and endsWith and optimize UTF8_BINARY_LCASE

### PR DESCRIPTION
**Pull request by vladimirg-db**
### What changes were proposed in this pull request?
Added hand-crafted implementations of unicode-aware lower-case `contains`, `startsWith`, `endsWith` to optimize UTF8_BINARY_LCASE for ASCII-only strings.

### Why are the changes needed?
`UTF8String.toLowerCase()`, which is used for the aforementioned collation-aware functions, has an optimization for full-ascii strings, but still always allocates a new object. In this PR I introduced loop-based implementations, which fall-back to `toLowerCase()` in case they meet a non-asci character.

### Does this PR introduce _any_ user-facing change?
No, these functions should behave exactly as:
- `lhs.containsInLowerCase(rhs)` == `lhs.toLowerCase().contains(rhs.toLowerCase())`
- `lhs.startsWithInLowerCase(rhs)` == `lhs.toLowerCase().startsWith(rhs.toLowerCase())`
- `lhs.endsWithInLowerCase(rhs)` == `lhs.toLowerCase().endsWith(rhs.toLowerCase())`

### How was this patch tested?
Added new test cases to `org.apache.spark.unsafe.types.CollationSupportSuite` and `org.apache.spark.unsafe.types.UTF8StringSuite`, including several unicode lowercase specific. Also I've run `CollationBenchmark` on GHA for JDK 17 and JDK 21 and have updated the data.

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Mind making the PR title complete? It's truncated.

**Comment by vladimirg-db** [comment]
This time I updated all the benchmarks results, since the JVM version (minor) on GHA workers has been bumped - otherwise results would be inconsistent and misleading.

**Comment by vladimirg-db** [comment]
@dbatomic, @uros-db, @cloud-fan, hi, requesting review

File: sql/core/benchmarks/CollationNonASCIIBenchmark-jdk21-results.txt [hunk 1]
```diff
@@ -1,54 +1,54 @@
-OpenJDK 64-Bit Server VM 21.0.2+13-LTS on Linux 6.5.0-1017-azure
+OpenJDK 64-Bit Server VM 21.0.3+9-LTS on Linux 6.5.0-1018-azure
 AMD EPYC 7763 64-Core Processor
 collation unit benchmarks - equalsFunction:  Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
 --------------------------------------------------------------------------------------------------------------------------
-UTF8_BINARY_LCASE                                   18244          18258          20          0.0      456096.4       1.0X
-UNICODE                                               498            498           0          0.1       12440.3      36.7X
-UTF8_BINARY                                           499            500           1          0.1       12467.7      36.6X
-UNICODE_CI                                          13429          13443          19          0.0      335725.4       1.4X
+UTF8_BINARY_LCASE                                   18412          18491         113          0.0      460288.0       1.0X
```
Comment (uros-db) [inline sql/core/benchmarks/CollationNonASCIIBenchmark-jdk21-results.txt:5]: is the performance for `UTF8_BINARY_LCASE` worse for all expressions (contains, startsWith, endsWith)?

File: sql/core/benchmarks/CollationNonASCIIBenchmark-jdk21-results.txt [hunk 2]
```diff
@@ -1,54 +1,54 @@
...
 ------------------------------------------------------------------------------------------------------------------------
-UTF8_BINARY_LCASE                                 23693          23695           3          0.0      592333.2       1.0X
-UNICODE                                            3170           3172           3          0.0       79243.5       7.5X
-UTF8_BINARY                                        3472           3473           2          0.0       86788.8       6.8X
-UNICODE_CI                                        63331          63603         384          0.0     1583274.3       0.4X
+UTF8_BINARY_LCASE                                 23384          23411          38          0.0      584606.2       1.0X
+UNICODE                                            2602           2603           0          0.0       65061.3       9.0X
+UTF8_BINARY                                        2901           2903           2          0.0       72536.6       8.1X
+UNICODE_CI                                       179352         179427         106          0.0     4483798.1       0.1X
```
Comment (vladimirg-db) [inline sql/core/benchmarks/CollationNonASCIIBenchmark-jdk21-results.txt:53]: I wonder what happened here...

File: sql/core/benchmarks/CollationNonASCIIBenchmark-jdk21-results.txt [hunk 3]
```diff
@@ -1,54 +1,54 @@
...
 ------------------------------------------------------------------------------------------------------------------------
-UTF8_BINARY_LCASE                                  9183           9230          67          0.0      229564.0       1.0X
-UNICODE                                           38937          38952          22          0.0      973421.3       0.2X
-UTF8_BINARY                                        1376           1376           0          0.0       34397.5       6.7X
-UNICODE_CI                                        32881          32882           1          0.0      822027.4       0.3X
+UTF8_BINARY_LCASE                                  9410           9413           5          0.0      235238.0       1.0X
+UNICODE                                           41038          41047          12          0.0     1025949.7       0.2X
+UTF8_BINARY                                        1406           1408           3          0.0       35151.5       6.7X
+UNICODE_CI                                        31829          31829           1          0.0      795717.5       0.3X
```
Comment (uros-db) [inline sql/core/benchmarks/CollationNonASCIIBenchmark-jdk21-results.txt:26]: I'd remove unrelated changes

File: sql/core/benchmarks/CollationBenchmark-results.txt [hunk 4]
```diff
@@ -1,54 +1,54 @@
...
 ------------------------------------------------------------------------------------------------------------------------
-UTF8_BINARY_LCASE                                 18110          18127          24          0.0      181103.9       1.0X
-UNICODE                                          171375         171435          85          0.0     1713752.3       0.1X
-UTF8_BINARY                                       14012          14030          26          0.0      140116.7       1.3X
-UNICODE_CI                                       153847         153901          76          0.0     1538471.1       0.1X
+UTF8_BINARY_LCASE                                 17963          17976          18          0.0      179630.9       1.0X
+UNICODE                                          171257         171257           1          0.0     1712565.3       0.1X
+UTF8_BINARY                                       10569          10574           7          0.0      105687.7       1.7X
+UNICODE_CI                                       149242         149281          55          0.0     1492422.8       0.1X
```
Comment (uros-db) [inline sql/core/benchmarks/CollationBenchmark-results.txt:26]: unrelated

File: sql/core/benchmarks/CollationBenchmark-results.txt [hunk 5]
```diff
@@ -1,54 +1,54 @@
...
-OpenJDK 64-Bit Server VM 17.0.10+7-LTS on Linux 6.5.0-1017-azure
+OpenJDK 64-Bit Server VM 17.0.11+9-LTS on Linux 6.5.0-1018-azure
 AMD EPYC 7763 64-Core Processor
 collation unit benchmarks - startsWith:   Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
 ------------------------------------------------------------------------------------------------------------------------
-UTF8_BINARY_LCASE                                 49520          49524           7          0.0      495195.4       1.0X
-UNICODE                                           18346          18346           0          0.0      183457.7       2.7X
-UTF8_BINARY                                       20483          20488           7          0.0      204828.7       2.4X
-UNICODE_CI                                       928756         930065        1851          0.0     9287564.4       0.1X
+UTF8_BINARY_LCASE                                 25579          25594          22          0.0      255786.4       1.0X
```
Comment (uros-db) [inline sql/core/benchmarks/CollationBenchmark-results.txt:41]: 50% improvement for startsWith/endsWith on ASCII, great!

File: sql/core/benchmarks/CollationNonASCIIBenchmark-jdk21-results.txt [hunk 6]
```diff
@@ -1,54 +1,54 @@
...
-OpenJDK 64-Bit Server VM 21.0.2+13-LTS on Linux 6.5.0-1017-azure
+OpenJDK 64-Bit Server VM 21.0.3+9-LTS on Linux 6.5.0-1018-azure
 AMD EPYC 7763 64-Core Processor
 collation unit benchmarks - contains:     Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
 ------------------------------------------------------------------------------------------------------------------------
-UTF8_BINARY_LCASE                                 22429          22438          13          0.0      560735.1       1.0X
-UNICODE                                            2900           2901           2          0.0       72503.2       7.7X
-UTF8_BINARY                                        3190           3198          11          0.0       79740.5       7.0X
-UNICODE_CI                                       166847         167278         609          0.0     4171180.3       0.1X
+UTF8_BINARY_LCASE                                 22537          22546          13          0.0      563430.8       1.0X
```
[the remark this answers is not in the document]
Comment (vladimirg-db) [reply sql/core/benchmarks/CollationNonASCIIBenchmark-jdk21-results.txt:32]: Not drastically (in the range of a measurement error I would say, judging by the other data). But it's expected with the current approach - we first have to find a non-ascii character, so part of the loop gives this slowdown.

File: sql/core/benchmarks/CollationBenchmark-jdk21-results.txt [hunk 7]
```diff
@@ -1,54 +1,54 @@
...
 ------------------------------------------------------------------------------------------------------------------------
-UTF8_BINARY_LCASE                                 11716          11716           1          0.0      117157.9       1.0X
-UNICODE                                          180133         180137           5          0.0     1801332.1       0.1X
-UTF8_BINARY                                       10476          10477           1          0.0      104757.4       1.1X
-UNICODE_CI                                       148171         148190          28          0.0     1481705.6       0.1X
+UTF8_BINARY_LCASE                                 11430          11437          10          0.0      114298.7       1.0X
+UNICODE                                          176891         176904          18          0.0     1768912.1       0.1X
+UTF8_BINARY                                        9898           9900           3          0.0       98977.9       1.2X
+UNICODE_CI                                       156862         156869          11          0.0     1568616.3       0.1X
```
Comment (uros-db) [inline sql/core/benchmarks/CollationBenchmark-jdk21-results.txt:26]: unrelated

Comment (vladimirg-db) [reply sql/core/benchmarks/CollationBenchmark-jdk21-results.txt:26]: Yea, I explained it in the uppermost comment before requesting a review - the JDK version on GHA workers has been bumped, so old resulds are no longer relevant to the "OpenJDK ..." header

File: sql/core/benchmarks/CollationBenchmark-results.txt (hunk 5 above)
Comment (uros-db) [reply sql/core/benchmarks/CollationBenchmark-results.txt:41]: just to check, is this benchmark for ASCII only?

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java
[the remark this answers is not in the document]
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:1]: please read more about preferred code structure here: https://issues.apache.org/jira/browse/SPARK-47410

File: common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java
Comment (uros-db) [inline common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java:1]: once new changes are moved to CollationSupport.CollationAwareUTF8String, then these tests can just end up in CollationSupportSuite

ref: https://issues.apache.org/jira/browse/SPARK-47410

File: common/unsafe/src/test/java/org/apache/spark/unsafe/types/UTF8StringSuite.java
Comment (uros-db) [inline common/unsafe/src/test/java/org/apache/spark/unsafe/types/UTF8StringSuite.java:1]: once new collatino-aware UTF8String changes are moved to CollationSupport.CollationAwareUTF8String, then these tests can just end up in CollationSupportSuite

ref: https://issues.apache.org/jira/browse/SPARK-47410

Comment (uros-db) [reply common/unsafe/src/test/java/org/apache/spark/unsafe/types/UTF8StringSuite.java:1]: it's okay to add more `UTF8_BINARY_LCASE` cases under assertContains/StartsWith/EndsWith to make sure everything works properly there

File: sql/core/benchmarks/CollationNonASCIIBenchmark-jdk21-results.txt (hunk 2 above)
Comment (vladimirg-db) [reply sql/core/benchmarks/CollationNonASCIIBenchmark-jdk21-results.txt:53]: Discussed this with @uris-db, performance has changed after this PR: https://issues.apache.org/jira/browse/SPARK-47863

File: sql/core/benchmarks/CollationBenchmark-jdk21-results.txt (hunk 7 above)
Comment (vladimirg-db) [reply sql/core/benchmarks/CollationBenchmark-jdk21-results.txt:26]: Discussed this on Slack - since the JDK version and some of the [numbers](https://github.com/apache/spark/pull/46082#pullrequestreview-2008030864) have changed, I will regenerate all the bench results in the separate PR.

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java
[the remark this answers is not in the document]
Comment (vladimirg-db) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:1]: Not possible, since the `getByte(...)` method is private (for a good reason in a utf-8 string)

File: common/unsafe/src/test/java/org/apache/spark/unsafe/types/UTF8StringSuite.java
Comment (vladimirg-db) [reply common/unsafe/src/test/java/org/apache/spark/unsafe/types/UTF8StringSuite.java:1]: Same as [here](https://github.com/apache/spark/pull/46082#discussion_r1570267391)

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 8]
```diff
@@ -359,10 +414,97 @@ public boolean startsWith(final UTF8String prefix) {
...
+    // No way to match sizes of strings for early return, since single grapheme can be expanded
+    // into several independent ones in lowercase
+    if (prefix.numBytes == 0) {
+      return true;
+    }
+    if (numBytes == 0) {
+      return false;
+    }
+
+    for (var i = 0; i < prefix.numBytes; i++) {
```
Comment (dbatomic) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:433]: Would it be possible to share implementation between startsWithInLowerCase and endsWithInLowerCase?

E.g. if you were to use two variables (offset in prefix and offset in this) you should be able to share the code between these two.

Comment (vladimirg-db) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:433]: I wanted to do something like this, but I think that this simple implementation is better for the next reasons:
- We couldn't do something like `matchAt(...)` by using an offset in `this`, because we actually don't know the exact offset (at least without the full conversion to lowercase), since lowercase representation of the unicode string can actually be longer (or shorter) than the original. That's why to check the prefix the loop is forward and to check the suffix the loop is backward
- We probably could do a parametrized function, which does something like `startOffset + this.offset + direction * i`, which would work for the loop indexing, but I can't think of the suitable math for the inner `if` from the top of my head. It's probably possible, but would be much more involved, since that code differs significantly in terms of indexing
- We could just substitute the loop body with the closure, and pass it as a strategy (since that's what differs), but introducing looping virtual call in an optimization PR wouldn't be reasonable
- It reads trivially, which is good for the reader
- The copy-paste here is not that large

**Review by dbatomic — APPROVED** [review]

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 9]
```diff
@@ -359,10 +414,97 @@ public boolean startsWith(final UTF8String prefix) {
...
+      final var lhsByte = getByte(i);
+      final var rhsByte = prefix.getByte(i);
+      if (lhsByte < 0 || rhsByte < 0) {
```
Comment (cloud-fan) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:436]: just for my curiosity: is it more efficient to do the ascii check ahead return earlier for the slow path? then we can make the fast path loop tighter.

**Comment by vladimirg-db** [comment]
Accidentally deleted the branch in my fork, opened a new PR: https://github.com/apache/spark/pull/46181
